rflx/expression.py

```diff
@@ -1078,6 +1120,9 @@ class UndefinedExpr(Name):
     def representation(self) -> str:
         return "__UNDEFINED__"
 
+    def __neg__(self) -> "UndefinedExpr":
+        return self.__class__(not self.negative, self.location)
+
```

Comment:
Why do we need to introduce negation for `UndefinedExpr`?

Reply:
`__neg__` is an abstract method of `Expr` so it has to be implemented.

Reply:
No, you made this abstract in the `Name` class in this PR (it's not abstract in `Expr`). I wonder whether it should either be abstract in `Expr` or not abstract at all. In any case, shouldn't `__neg__` for `UndefinedExpr` just raise `NotImplementedError` (you're not supposed to use or negate it...)?

Reply:
It is abstract in `Expr`. The abstract definition in `Name` is redundant and I will remove it. Also we need an implementation because we often create and evaluate expressions that might (or might not) contain undefined expressions. I don't think we should throw an exception if this is the case.

Reply:
Wouldn't it be faster to ensure that an expression does not contain `UNDEFINED` before executing a simplification or substitution?

Reply:
Here are the necessary changes:

```diff
diff --git a/rflx/expression.py b/rflx/expression.py
index 0172c4bd..4f5c3eff 100644
--- a/rflx/expression.py
+++ b/rflx/expression.py
@@ -1121,7 +1121,7 @@ class UndefinedExpr(Name):
         return "__UNDEFINED__"
 
     def __neg__(self) -> "UndefinedExpr":
-        return self.__class__(not self.negative, self.location)
+        raise NotImplementedError
 
     def z3expr(self) -> z3.ExprRef:
         raise NotImplementedError
diff --git a/rflx/pyrflx/typevalue.py b/rflx/pyrflx/typevalue.py
index 8513ec0f..00d44e9d 100644
--- a/rflx/pyrflx/typevalue.py
+++ b/rflx/pyrflx/typevalue.py
@@ -595,6 +595,8 @@ class MessageValue(TypeValue):
                 return first if isinstance(first, Number) else None
         prv = self._prev_field(fld)
         if prv:
+            if UNDEFINED in (self._fields[prv].first, self._fields[prv].typeval.size):
+                return None
             first = self.__simplified(Add(self._fields[prv].first, self._fields[prv].typeval.size))
             return first if isinstance(first, Number) else None
         return None
@@ -1056,6 +1058,8 @@ class MessageValue(TypeValue):
             self.name_length = Length(name)
 
         def _last(self) -> Expr:
+            if self.first == UNDEFINED:
+                return UNDEFINED
             return Sub(Add(self.__first, self.typeval.size), Number(1)).simplified()
 
         @property
```

Reply:
> Wouldn't it be faster to ensure that an expression does not contain UNDEFINED before executing a simplification or substitution?

I thought so, too. Interestingly this does not seem to be the case and we lose 3-5% performance.

Reply:
Could it be caused by the duplicate dictionary accesses to `self._fields`?

Reply:
It is probably that and the call to `Field.first`/`Field.size`. The performance impact seems to be smaller than I initially thought. While it is consistently slower than without the change it's much less than 5% after multiple runs. Since we're going to use a different approach for the desired performance in the fuzzer anyway I'd just leave it as is.

Reply:
Fixed in https://github.com/Componolit/RecordFlux/pull/390/commits/2cf704854837207e59fccc16368733c42d6e0890.